pkg/addonmanager/controllers/agentdeploy/utils.go

```diff
@@ -216,63 +223,100 @@ func (m *managedManifest) deployable(hostedModeEnabled bool, installMode string,
+		deployable, err := b.processor.deployable(b.hostedModeEnabled, installMode, object)
 		if err != nil {
 			return nil, nil, err
 		}
+		if !deployable {
+			continue
+		}
+
+		isHookObject, _ := b.isPreDeleteHookObject(object)
+		if isHookObject {
+			continue
+		}
+		deployObjects = append(deployObjects, object)
+	}
+	if len(deployObjects) == 0 {
+		return nil, nil, nil
+	}
+
+	return b.workBuilder.Build(deployObjects,
+		newAddonWorkObjectMeta(b.processor.manifestWorkNamePrefix(addon.Namespace, addon.Name), addon.Name, addon.Namespace, addonWorkNamespace, owner),
+		workbuilder.ExistingManifestWorksOption(existingWorks),
+		workbuilder.ManifestConfigOption(manifestOptions))
+}
+
+// BuildHookWork returns the preDelete manifestWork, if there is no manifest need
+// to deploy, will return nil.
+func (b *addonWorksBuilder) BuildHookWork(addonWorkNamespace string,
+	addon *addonapiv1alpha1.ManagedClusterAddOn,
+	objects []runtime.Object) (hookWork *workapiv1.ManifestWork, err error) {
+	var hookManifests []workapiv1.Manifest
+	var hookManifestConfigs []workapiv1.ManifestConfigOption
+	var owner *metav1.OwnerReference
+	installMode, _ := constants.GetHostedModeInfo(addon.GetAnnotations())
+	if installMode == constants.InstallModeDefault {
+		owner = metav1.NewControllerRef(addon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
+	}
 
+	for _, object := range objects {
 		deployable, err := b.processor.deployable(b.hostedModeEnabled, installMode, object)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		if !deployable {
 			continue
 		}
 
 		isHookObject, manifestConfig := b.isPreDeleteHookObject(object)
-		if isHookObject {
-			hookManifests = append(hookManifests, workapiv1.Manifest{
-				RawExtension: runtime.RawExtension{Raw: rawObject},
-			})
-			manifestConfigs = append(manifestConfigs, *manifestConfig)
-		} else {
-			deployManifests = append(deployManifests, workapiv1.Manifest{
-				RawExtension: runtime.RawExtension{Raw: rawObject},
-			})
+		if !isHookObject {
+			continue
+		}
+		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+		if err != nil {
+			return nil, err
 		}
-	}
 
-	deployManifestWork = newManifestWork(addon.Namespace, addonName, manifestWorkNamespace,
-		deployManifests, b.processor.manifestWorkName)
-	hookManifestWork = newManifestWork(addon.Namespace, addonName, manifestWorkNamespace,
-		hookManifests, b.processor.preDeleteHookManifestWorkName)
-	if hookManifestWork != nil {
-		hookManifestWork.Spec.ManifestConfigs = manifestConfigs
+		hookManifests = append(hookManifests, workapiv1.Manifest{RawExtension: runtime.RawExtension{Raw: rawObject}})
+		hookManifestConfigs = append(hookManifestConfigs, *manifestConfig)
+	}
+	if len(hookManifests) == 0 {
+		return nil, nil
 	}
 
-	return deployManifestWork, hookManifestWork, nil
+	hookWork = newManifestWork(addon.Namespace, addon.Name, addonWorkNamespace, hookManifests, b.processor.preDeleteHookManifestWorkName)
```

Comment:
can we use the same build func?

Reply:
yes, but it will not be easy to understand the returned work slices since we only define only a manifestWork for the pre-delete hook manifests, and do not need existing works to build.